tests/callbacks/test_RecorderListener.cpp

```diff
@@ -0,0 +1,159 @@
+#include "gtest/gtest.h"
+
+#include "Kokkos_Core.hpp"
+
+#include "kokkos-utils/impl/type_traits.hpp"
+
+#include "kokkos-utils/callbacks/EventInProfileSectionRegexMatcher.hpp"
+#include "kokkos-utils/callbacks/EventRegexMatcher.hpp"
+#include "kokkos-utils/callbacks/Manager.hpp"
+#include "kokkos-utils/callbacks/RecorderListener.hpp"
+
+#include "tests/callbacks/Helpers.hpp"
+#include "tests/callbacks/TestWorkload.hpp"
+
+/**
+ * @addtogroup unittests
+ *
+ * @c Kokkos callback recorder listener
+ * ------------------------------------
+ *
+ * This group of tests check the behavior of @ref Kokkos::utils::callbacks::RecorderListener.
+ */
+
+using execution_space = Kokkos::DefaultExecutionSpace;
+
+namespace Kokkos::utils::tests::callbacks
+{
+
+using namespace Kokkos::utils::callbacks;
+
+//! Listener to record events that occur in a profile section.
+using event_in_profile_section_recorder_t = RecorderListener<EventInProfileSectionRegexMatcher, EventTypeList>;
+
+class RecorderListenerTest : public ManagerTestFixture
+{
+public:
+    void SetUp() override {
+        this->exec = Kokkos::Experimental::partition_space(execution_space{}, 1)[0];
+    }
+
+protected:
+    execution_space exec {};
+};
+
+//! @test Check traits of @ref Kokkos::utils::callbacks::RecorderListener.
+TEST_F(RecorderListenerTest, traits)
+{
+    //! Check that @ref event_in_profile_section_recorder_t is a listener for any event.
+    static_assert(std::same_as<
+        listener_event_type_list_t<event_in_profile_section_recorder_t>,
+        EventTypeList
+    >);
+
+    static_assert(Listener<event_in_profile_section_recorder_t>);
+}
+
+//! @test Check the behavior of @ref Kokkos::utils::callbacks::RecorderListener.
+TEST_F(RecorderListenerTest, recorded_events)
+{
+    EventInProfileSectionRegexMatcher matcher(std::regex("profile section"));
+    const auto recorder = std::make_shared<event_in_profile_section_recorder_t>(std::move(matcher));
+
+    const auto any_event_recorder = std::make_shared<RecorderListener<EventTypeList>>();
+
+    Manager::register_listener(recorder);
+    Manager::register_listener(any_event_recorder);
+
+    MyWorkload<execution_space>{}.execute(exec);
+
+    //! Check the recorded events.
+    ASSERT_THAT(
+        recorder->recorded_events,
+        ::testing::Contains(BeginParallelForEventWithName(::testing::StrEq("computation - level 0 - pfor")))
+    );
+
+    ASSERT_THAT(
+        recorder->recorded_events,
+        ::testing::Contains(BeginFenceEventWithName(::testing::StrEq("computation - level 0 - fence after pfor")))
+    );
+
+    ASSERT_THAT(
+        recorder->recorded_events,
+        ::testing::Contains(ProfileEventWithName(::testing::StrEq("buried marker")))
+    );
+
+    ASSERT_THAT(
+        recorder->recorded_events,
+        ContainsInOrder(
+            BeginParallelForEventWithName(::testing::StrEq("computation - level 0 - pfor")),
+            ProfileEventWithName(::testing::StrEq("buried marker"))
+        )
```

Comment:
Put them all.

Reply:
(I mean all the events that we see from `TestWorkload.hpp`.)

Reply:
OK. Was some work :) But it's done now! Ouf. I noted that we have "tool fences". I didn't put those in the matcher. We could see how to disable those tool fences.